File: 백준/Gold/1759. 암호 만들기/암호 만들기.java


import java.util.*;
import java.io.*;

public class Main {

    static int L, C;
    static String[] alp;
    static ArrayList<String> result = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        L = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        alp = new String[C];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < C; i++) {
            alp[i] = st.nextToken();
        }

        Arrays.sort(alp);

        backtrack(0, 0, new StringBuilder());

        for (String res : result) {
            System.out.println(res);
        }
    }

    static void backtrack(int dep, int pre, StringBuilder password) {
        if (dep == L) {
            int vowelCount = 0;
            int consonantCount = 0;

            for (int i = 0; i < password.length(); i++) {
                char ch = password.charAt(i);
                if (isVowel(ch)) {
                    vowelCount++;
                } else {
                    consonantCount++;
                }
            }

            // 모음이 1개 이상이고 자음이 2개 이상인 경우만 추가
            if (vowelCount >= 1 && consonantCount >= 2) {
                result.add(password.toString());
            }
            return;
        }

        for (int i = pre; i < C; i++) {
            password.append(alp[i]);
            backtrack(dep + 1, i + 1, password);
            password.deleteCharAt(password.length() - 1); // 백트래킹
        }
    }

    static boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
}

File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
# import sys
# sys.stdin = open("input.txt", "r")

L, C = map(int, input().split())
arr = input().split()
arr.sort()  # 사전순 정렬

answer = []
result = []

# 모음과 자음 조건 확인
def check(temp):
    mo = 0  # 모음 개수
    ja = 0  # 자음 개수

    for t in temp:
        if t in ('a', 'e', 'i', 'o', 'u'):
            mo += 1
        else:
            ja += 1

    return mo >= 1 and ja >= 2  # 최소 모음 1개, 자음 2개 조건

# 백트래킹 함수
def backtrack(dep, pre):
    if dep == L:  # 길이가 L인 암호 완성
        if check(answer):  # 조건에 맞는 경우만 결과에 추가
            result.append("".join(answer))
        return

    for i in range(pre, C):
        answer.append(arr[i])
        backtrack(dep + 1, i + 1)
        answer.pop()

# 백트래킹 시작
backtrack(0, 0)

# 결과 출력
for res in result:
    print(res)
